x/expiration/keeper/keeper.go

```diff
@@ -0,0 +1,565 @@
+	}
+
+	// attempt to send coins from owner account to expiration module account
+	depErr := k.bankKeeper.SendCoinsFromAccountToModule(ctx, ownerAddr, types.ModuleName, expiration.Deposit)
+	if depErr != nil {
+		return types.ErrInsufficientDeposit.Wrap(depErr.Error())
+	}
+
+	// emit event indicating that a deposit was collected
+	k.emitEvent(ctx, types.NewEventExpirationDeposit(expiration.ModuleAssetId, expiration.Owner, expiration.Deposit))
+	if err != nil {
+		return types.ErrSetExpiration.Wrapf(
+			"failed to emit EventExpirationDeposit for [%s]: %v", expiration.Owner, err)
+	}
+
+	// marshal expiration record and store
+	b, err := k.cdc.Marshal(&expiration)
+	if err != nil {
+		return types.ErrSetExpiration.Wrap(
+			types.ErrUnmarshal.Wrapf("expiration: %v: %v", expiration, err).Error())
+	}
+	store.Set(key, b)
+
+	// emit event indicating the expiration was added
+	k.emitEvent(ctx, types.NewEventExpirationAdd(expiration.ModuleAssetId))
+
+	return nil
+}
+
+// ExtendExpiration extends an expiration to time in the future.
+func (k Keeper) ExtendExpiration(ctx sdk.Context, expiration types.Expiration) error {
+	// get key
+	key, err := types.GetModuleAssetKey(expiration.ModuleAssetId)
+	if err != nil {
+		return types.ErrExtendExpiration.Wrapf("%s [%s]: %v",
+			types.ErrInvalidKey.Error(), expiration.ModuleAssetId, err)
+	}
+
+	// lookup old expiration
+	oldExpiration, err := k.GetExpiration(ctx, expiration.ModuleAssetId)
+	if err != nil {
+		return types.ErrExtendExpiration.Wrap(err.Error())
+	}
+	if oldExpiration == nil {
+		return types.ErrExtendExpiration.Wrapf("%s: %s", types.ErrNotFound.Error(), expiration.ModuleAssetId)
+	}
+
+	// Validate expiration time is after old expiration time
+	if oldExpiration.Time.After(expiration.Time) {
+		return types.ErrExtendExpiration.Wrapf("new expiration time [%s] must be after old expiration time [%s]",
+			expiration.Time.UTC(), oldExpiration.Time.UTC())
+	}
+	// Validate owners are the same
+	if expiration.Owner != oldExpiration.Owner {
+		return types.ErrNewOwnerNoMatch.Wrapf("new owner [%s] and old owner [%s] do not match",
+			expiration.Owner, oldExpiration.Owner)
+	}
+
+	// Marshal expiration record and store
+	store := ctx.KVStore(k.storeKey)
+	b, err := k.cdc.Marshal(&expiration)
+	if err != nil {
+		return types.ErrExtendExpiration.Wrapf("%s: %v", types.ErrUnmarshal.Error(), err)
+	}
+	store.Set(key, b)
+
+	// emit Extend event
+	k.emitEvent(ctx, types.NewEventExpirationExtend(expiration.ModuleAssetId))
+
+	return nil
+}
+
+// InvokeExpiration invokes an expiration message through the MsgServiceRouter.
+// The expiration message is removed when the message invocation is successful.
+//
+// Caution: Do NOT call InvokeExpiration from within a module to delete a self-owned asset.
+// This will lead to an infinite loop as expiration messages in this method are routed through
+// the message router. Use RemoveExpiration instead.
```

Comment:
It'd probably be worthwhile to create an `exported/exported.go` with an interface that this keeper fulfills, but only has the functions we expect people to want in other modules, e.g. `SetExpiration` and `RemoveExpiration` (and maybe `GetDeposit` and `GetDuration` depending on other discussions).

Modules that want to implement expirations could then import that for the field in their keeper. That would dissuade use of this function and also cut down on circular dependency possibilities.